Modules/posixmodule.c

```
@@ -8652,12 +8651,20 @@ posix_sendfile(PyObject *self, PyObject *args, PyObject *kwdict)
             }
             if (i > 0) {
                 sf.hdr_cnt = (int)i;
-                i = iov_setup(&(sf.headers), &hbuf,
-                              headers, sf.hdr_cnt, PyBUF_SIMPLE);
-                if (i < 0)
+                if (iov_setup(&(sf.headers), &hbuf,
+                              headers, sf.hdr_cnt, PyBUF_SIMPLE) < 0)
                     return NULL;
 #ifdef __APPLE__
-                sbytes += i;
+                for (i = 0; i < sf.hdr_cnt; i++) {
+                    Py_ssize_t blen = sf.headers[i].iov_len;
+# define OFF_T_MAX 0x7fffffffffffffff
```

I'm not sure about this. Is `off_t` always 64-bit on macOS?

---

According to `pymacconfig.h`, `off_t` is equal to `long long` on 32-bit, and `long` on 64-bit. E.g. it is always 64-bit.

---

I've checked the SDK headers to be sure, and off_t is indeed always 64-bit (off_t is a typedef for __darwin_off_t, and that in turn is a typedef for __int64_t). 

The manpage for sendfile(2) doesn't mention this, but the one for writev(2) says:

     [EINVAL]           The sum of the iov_len values in the iov array overflows a 32-bit integer.

Furthermore we have had problems with write(2) in the past when writing more than INT_MAX bytes at a time (see the implementation of _Py_write_impl).  I think it is therefore better to limit the maximum size to INT_MAX, unless we explicitly test that sendfile can send more than INT_MAX bytes in one go.

BTW. Instead of OFF_T_MAX you could use OFF_MAX here, that should be defined to the maximum value of off_t.

---

Is OFF_MAX a macOS specific constant?

---

We have not just the sum of the iov_len values, but also added the length of the data sent from a file. The total sum can exceed 32-bit limit even if the sum of the iov_len values is in 32-bit limit.

I think it is better to use a 64-bit limit and allow OS to return an error if this is not supported. We should guard only from integer overflows occurred on the border between Python and OS.

---

In _PyWrite_impl we explicitly guard against overflowing a 32-bit integer to avoid problems. IIRC this was done because this gives a nicer user experience (and hoping that Apple will fix this in the kernel is not really useful).

I'll see if I can create a simple test to check if sendfile suffers from the same limitation on 10.13 and 10.14.

---

In _PyWrite_impl there is a different situation. We guard against overflowing `int` on Windows because `write()` accepts only `unsigned int` as the *count* argument on Windows, and because it is allowed to write less bytes then requested in `write()`. This is a border between Python and OS.

`readv()`, `writev()` and `sendfile()` accept `size_t` as the length of every chunk on all supported OSes even if OS doesn't support writing too large total size. I'm not sure that it is better to silently truncate the output than raise OSError. At least `sendfile()` with *headers* and *trailers* will produce corrupted output in this case: headers, than truncated content, than trailers.

---

_Py_write_impl does indeed not limit writes, I was confused because I have a local patch for [issue 24658](https://bugs.python.org/issue24658) that I never pushed... 

sendfile like write can write less bytes than requested (at least, that's what the API suggests as it returns the amount of bytes written), limiting the amount of data written to avoid misbehaviour of the OS shouldn't be a problem.

That said: limiting is only necessary when there is a reason to do so.  I guess I'll really have to write a test program that checks what the system does when trying to write more than 2GB of data.

---

Please test sending 2 GiB with headers and trailers.
```python
os.sendfile(sockno, fileno, 0, 2**31-1, headers=[b"[begin]"], trailers=[b"[end]"])
```
If truncate the total size to INT_MAX, then this can send `b"[begin]"`, then `2**31-1-7` bytes from the file (instead of expected `2**31-1` bytes) and then `b"[end]"`. This is a wrong result.